chained/water_impl.go

```diff
@@ -23,85 +23,107 @@ type waterImpl struct {
-		}
-		d.dialer = dialer
+			d.dialer, err = createDialer(ctx, wasm, transport)
+			if err != nil {
+				d.errLoadingWASM = log.Errorf("failed to create dialer: %w", err)
+				return
+			}
+			d.setReady()
+		}()
 		return d, nil
 	}
 
 	if wasmAvailableAt != "" {
-		wg.Add(1)
 		go func() {
-			defer wg.Done()
-			log.Debugf("Loading WASM for %q. If not available, it should try to download from the following URLs: %+v. The file should be available here: %s", transport, strings.Split(wasmAvailableAt, ","), dir)
-			vc := newWaterVersionControl(dir)
-			cli := httpClient
-			if cli == nil {
-				cli = proxied.ChainedThenDirectThenFrontedClient(1*time.Minute, "")
-			}
-			downloader, err := newWaterWASMDownloader(strings.Split(wasmAvailableAt, ","), cli)
-			if err != nil {
-				d.downloadErr = log.Errorf("failed to create wasm downloader: %w", err)
-				return
-			}
+			defer d.finishedLoading()
+			log.Debugf("Loading WASM for %q. If not available, it should try to download from the following URLs: %+v. The file should be available at: %q", transport, strings.Split(wasmAvailableAt, ","), dir)
 
-			r, err := vc.GetWASM(ctx, transport, downloader)
+			r, err := d.loadWASM(ctx, transport, dir, wasmAvailableAt)
 			if err != nil {
-				d.downloadErr = log.Errorf("failed to get wasm: %w", err)
+				d.errLoadingWASM = log.Errorf("failed to read wasm: %w", err)
 				return
 			}
 			defer r.Close()
-
 			b, err := io.ReadAll(r)
 			if err != nil {
-				d.downloadErr = log.Errorf("failed to read wasm: %w", err)
+				d.errLoadingWASM = log.Errorf("failed to load wasm bytes: %w", err)
 				return
 			}
 
-			if len(b) == 0 {
-				d.downloadErr = log.Errorf("received empty wasm")
-				return
-			}
+			log.Debugf("received wasm with %d bytes", len(b))
 
-			dialer, err := createDialer(ctx, b, transport)
+			d.dialer, err = createDialer(ctx, b, transport)
 			if err != nil {
-				d.downloadErr = log.Errorf("failed to create dialer: %w", err)
+				d.errLoadingWASM = log.Errorf("failed to create dialer: %w", err)
+				return
 			}
-			d.dialer = dialer
+			d.setReady()
 		}()
 	}
 
 	return d, nil
 }
 
+func (d *waterImpl) loadWASM(ctx context.Context, transport string, dir string, wasmAvailableAt string) (io.ReadCloser, error) {
+	waterLoadingWASMMutex.Lock()
+	defer waterLoadingWASMMutex.Unlock()
+	vc := newWaterVersionControl(dir)
+	cli := httpClient
+	if cli == nil {
+		cli = proxied.ChainedThenDirectThenFrontedClient(1*time.Minute, "")
+	}
+	downloader, err := newWaterWASMDownloader(strings.Split(wasmAvailableAt, ","), cli)
+	if err != nil {
+		return nil, log.Errorf("failed to create wasm downloader: %w", err)
+	}
+	r, err := vc.GetWASM(ctx, transport, downloader)
+	if err != nil {
+		return nil, log.Errorf("failed to get wasm: %w", err)
+	}
+	return r, nil
+}
+
+func (d *waterImpl) finishedLoading() {
+	select {
+	case d.finishedToLoad <- struct{}{}:
+	default:
+		log.Error("channel closed")
+	}
```

Comment:
There is no safe way to write to a channel if you don't know whether it's closed.  This call should only happen if we are sure the channel is open.  To avoid this dilemma, only the goroutine responsible for writing to a channel should close that channel (if it needs to).

Here's an example of this kind of `select` statement resulting in a panic: https://go.dev/play/p/SUgKWfok-1Y